### PR TITLE
Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Alfredbis29**
## Hello 👋 Reviewer,
In this project, my partner and I have added validations to the models, created specs for them,  found and fixed N+1 query problems.

### TASKS COMPLETED
Add the following validations:
> For the User model:

- Name must not be blank.
- PostsCounter must be an integer greater than or equal to zero.
> For the Post model:

- Title must not be blank.
- Title must not exceed 250 characters.
- CommentsCounter must be an integer greater than or equal to zero.
- LikesCounter must be an integer greater than or equal to zero.
- Add unit specs for all of your models' methods and validations.
- Add flash messages in the create actions in all your controllers.
- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. This is the corresponding wireframe

### CODE REVIEW

**Review by TedLivist — COMMENTED** [review]
## STATUS: INVALID CODE REVIEW REQUEST 🛑 🛑

Hi @Alfredbis29 and @keza681,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

### You have submitted a project with failed Rubocop linters

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Linda, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
- [x] your project should have featured tests for User index page:
- Tests that assert you can see the username of all other users.
- Tests that assert you can see the profile picture for each user.
- Tests that assert you can see the number of posts each user has written.
- Tests that redirect to the user's show page when you click on a user.

- [x] your project should have featured tests for User show page:
- Tests that asserts you can see the user's profile picture.
- Tests that asserts you can see the user's username.
- Tests that asserts you can see the number of posts the user has written.
- Tests that asserts you can see the user's bio.
- Tests that asserts you can see the user's first 3 posts.
- Tests that asserts you can see a button that lets me view all of a user's posts.
- Tests that redirects to the post's show page when you click on a user's post.
- Tests that redirects to the user's post's index page when you click on a see all posts.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Alfredbis29** [comment]
Hi there, 
Thanks for the professional review @mikethreels  , We have done all requirement as mention ,
you can now check thanks.

**Review by arslanbisharat — APPROVED** [review]
Hi Team,
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- The Pull Request has a proper title and description. ✔️ 
- You have added a descriptive Readme file. Good Job. ✔️ 
- There is no linter error is present. ✔️ 
- Project has met all the technical aspects you needed for this project. Keep rocking! ✔️ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integration/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      @third_user.save!
+    end
+
+    it 'shows the users username' do
+      visit root_path
+      expect(page).to have_content('Mohammed')
+      expect(page).to have_content('Ahmed')
+      expect(page).to have_content('Marwan')
+    end
+
```
Comment (brytebee) [inline spec/integration/user_index_spec.rb:20]: - [x] **Required:** You are required to have a test for a link to the show page. Kindly include it.

File: spec/integration/post_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      @comment1 = Comment.create(post: Post.first, author: User.first, text: 'Hi Tom!')
+      @comment2 = Comment.create(post: Post.first, author: User.first, text: 'Hi Tom Again!')
+      @comment3 = Comment.create(post: Post.first, author: User.first, text: 'Hi Tom Again twice!')
+
+      visit user_post_path(@first_user, @first_post)
+    end
+
+    it 'shows posts title' do
+      expect(page).to have_content('Hello')
+    end
```
Comment (brytebee) [inline spec/integration/post_show_spec.rb:30]: - [x] **Required:** This test fails. Kindly fix it.

image

**Review by brytebee — CHANGES_REQUESTED** [review]
Hi, @Alfredbis29 & @keza681 👋,

Good job so far coming so far with the progress you've made! 💪💪

## Highlights
You have done amazing in the following:
- The linters are correctly installed. All green!✔️
- The rails gem is installed and the request type is used.✔️
 
However, there are some issues that you still need to work on to go to the next project but you are almost there!



# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Please, remember to tag me in your question so I can receive the notification.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @brytebee in your question so I can receive the notification.
### _**Use @brytebee on slack for faster responses.**_


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integration/post_show_spec.rb (hunk 2 above)
Comment (keza681) [reply spec/integration/post_show_spec.rb:30]: Hello @brytebee, Actually I have resolved the issue, and all the tests are passing well. Below is the screenshot

Screenshot 2022-07-26 at 01 57 27

**Review by ShoiraTa — APPROVED** [review]
## **STATUS: APPROVED  🟢 ** 
------
### Hi @Alfredbis29 , 

**Great work on making the changes requested by a previous reviewer 👏🏻** 

Your project is complete! . There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

5SW

**Cheers and Happy coding! :clap: :clap: :clap:**

Feel free to leave any questions or comments in the PR thread by mentioning my Github username if something is not 100% clear.
***
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have  limited number of  reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**